Audio buffer support in filter graph

egao1980 opened the pull request.

mikeboers commented:
This is looking quite lovely! Thank you.

[This line](https://github.com/mikeboers/PyAV/blob/bb9600ff8597c0ff3f35bb640be2e8fe3d297c25/av/audio/frame.pyx#L126) lies, but that is easy to change or add that feature like the `VideoFrame` variant.

My only lingering feeling is: since the planar audio comes back as 2D, should the non-planar audio do the same (except the axis be reversed)?

jlaine commented on av/filter/graph.pyx:
```diff
@@ -143,10 +145,46 @@ cdef class Graph(object):
 
         return self.add('buffer', args, name=name)
 
+    def add_abuffer(self, template=None, sample_rate=None, format=None, layout=None, channels=None, name=None):
```
I haven't had a chance to review this actual pice of code, but in any case it deserves plenty of unit tests, especially considering all the `if` branches which raise exceptions: these need to be tested with assertRaises

mikeboers replied:
... or at the minimum a couple of integration tests. At least asserting that we haven't broken it horribly in the future.

egao1980 replied:
Sure, I'll add some tests. It might take a while though. Most likely will port C audio filtering examples.

egao1980 replied:
Added a filter_audio.py - ported a C demo. Touches most of the initialization logic.

mikeboers commented on setup.py:
```diff
@@ -289,8 +289,10 @@ def compile_check(code, name, includes=None, include_dirs=None, libraries=None,
         objects = cc.compile([source_path], include_dirs=include_dirs)
         if link:
-            cc.link_executable(objects, exec_path, libraries=libraries,
-                library_dirs=library_dirs, runtime_library_dirs=runtime_library_dirs)
```
Looks like a common formatting disagreement. Really makes me want to adopt a strict auto-formatter.

(The problem there is that I personally tend to disagree with more formatters. I should stop that.)

jlaine replied:
I used to have my own opinions on formatting but switched to full-PEP8, it's nice to have consistency and you can expect contributors to be familiar with it

mikeboers replied:
Having quickly checked the first few examples of <https://www.python.org/dev/peps/pep-0008/?#indentation> I conclude that:

- What was removed was not PEP8-y (due to following lines not aligning).
- What was added was PEP8-y (due to them aligning).
- What I would naturally do is also PEP8-y (hanging indents):

```
cc.link_executable(
    objects,
    exec_path,
    # etc.
)
```

jlaine replied:
I can confirm that flake8 supports .pyx files, all you need is to add the following to setup.cfg:

```ini
[flake8]
filename=*.py,*.pyx,*.pxd
```

Then you can run `flake8 av`

egao1980 replied:
I've fixed indentation in the code fragment - we can have a separate issue to reformat the code according to PEP8

jlaine replied:
Could you please check whether this is still needed now that we don't do feature detection anymore?

egao1980 replied:
Sure.

mikeboers commented on av/filter/graph.pyx:
```diff
@@ -5,6 +5,8 @@ from av.filter.filter cimport Filter, wrap_filter
 from av.utils cimport err_check
 from av.video.frame cimport VideoFrame, alloc_video_frame
 from av.video.format cimport VideoFormat
+from av.audio.frame cimport AudioFrame, alloc_audio_frame
```
Agreed, but minor.

jlaine replied:
This could be enforced in CI using `isort` , provided it understands .pyx

mikeboers commented on av/audio/frame.pyx:
```diff
@@ -130,8 +131,16 @@ cdef class AudioFrame(Frame):
+                's32p':'<i4',
+                'flt' :'<f4',
                 'fltp':'<f4',
+                'dbl' :'<f8',
```
I haven't really dug into it, but I suspect that they can because of the numpy stacking thing. I did have questions there about adopting a strict `shape` protocol.

E.g. if we have 100 audio samples in stereo, planar would have `shape == (2, 100)`, but would non-planar have `shape == (100, 2)` or `(200, )`?

I feel like it should be `(100, 2)`; it is always 2-D, and the channel layout is always expressed in the array shape.

jlaine replied:
This has been dealt with in a more thorough fashion in #422, including the reciprocal `from_ndarray`

egao1980 commented:
I've added a ported filter_audio example. The filter chain looks usable now.  Please review.

jlaine commented:
I'm not familiar with the libav terminology but what is "abuffer"? Unless this is the exact name of something in libav, "audio buffer" would be more explicit

mikeboers commented:
@jlaine Unfortunately, [it really is called "abuffer"](https://ffmpeg.org/ffmpeg-filters.html#abuffer).

egao1980 commented:
It looks like appveyor 3.5 build is failing on random dependency download error. Is it possible to trigger rebuild?

egao1980 commented:
Could someone please re-review this one so it could be merged.

jlaine commented:
I think thusnis very cool, but as there are zero tests for this new code, I'm not comfortable with merging it. Doubly so as there are a *lot* of conditional branches..

egao1980 commented on examples/filter_audio.py:
```diff
@@ -0,0 +1,129 @@
+
+
+if __name__ == '__main__':
+    if len(sys.argv) < 2:
```
This example is as close as possible to the original FFMpeg demo. I can make the duration argument optional by choosing sensible default value.

egao1980 replied:
I've changed the example to run 1 second sample by default

egao1980 commented on av/filter/graph.pyx:
```diff
@@ -143,10 +146,50 @@ cdef class Graph(object):
 
+    def add_abuffer(self, template=None, sample_rate=None, format=None, layout=None, channels=None, name=None, time_base=None):
+        """
+        Convenient method for adding `abuffer <https://ffmpeg.org/ffmpeg-filters.html#abuffer>`_.
```
Thanks, done

egao1980 commented:
Hi Jeremy, the Graph part is currently undertested anyway so I'll just add a few tests for audio filters.

jlaine commented:
Right the current filter graph code is untested, otherwise the "elif" error would have been caught. We are currently trying to raise the test coverage so any new code must come with tests.

morrolinux commented:
Any updates on this? 
Since I really need this feature I'm currently using this branch from @egao1980 's Fork. working flawlessly for now

egao1980 commented:
Hi @morrolinux, I'm currently lack free time to write proper set of unit tests for this one. I'll fix the conflicts though.

mikeboers commented:
Thanks @egao1980 for trying to do that. My apologies for rarely having free time either; paying the bills has to come first, unfortunately.

tetelestia commented:
Just bumping this up, hoping it could be merged in. I've been using this branch for a while with no problems.

egao1980 commented:
https://github.com/TUW-GEO/pytesmo/issues/177

egao1980 commented:
Please re-review as I've added error codes, fixed example, and most importantly - added a few unit tests covering audio filters. 

Please note that several Travis configurations fail to initialise even before actually building the project.

mikeboers commented:
I'm interested in working through this, but it really turned into a monolith.

Anything that isn't related to the specific purpose of this has to be dealt with elsewhere (e.g. build and test fixes). The error codes stuff sits particularly badly with me, as it stretches so much further than audio buffers and was complete unnecessary AFAICT.

If someone wants to champion this, it will need to be based off of the current develop (not with the current develop merged into it!), and be isolated to the audio buffer support.

I'll try to take a look at it myself, as I'm on a bit of a PyAV binge, but I make no promises. I honestly don't know the scope of it because the history as-is is way too convoluted. If I do it, I might start from scratch and lose the authorship information.

mikeboers commented:
Closing in favour of #562.
